### PR TITLE
Owasm executor on google cloud function

### LINKED ISSUE #1825: Setup google cloud service to setup cloud execution

### PR DESCRIPTION
**Pull request by perimeko**
Fixed #1825

### CODE REVIEW

File: cloud-function/google_cloud_function.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+    path = "/tmp/execute.sh"
+    with open(path, "w") as f:
+        f.write(request["executable"])
+
+    os.chmod(path, 0o775)
+
+    try:
+
+        timeout_millisec = request_json["timeout"]
```
Comment (sorawit) [inline cloud-function/google_cloud_function.py]: Can't we use `request.timeout`?

File: cloud-function/google_cloud_function.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        if type(timeout_millisec) == str:
+            timeout_millisec = int(timeout_millisec)
+        timeout_sec = timeout_millisec / 1000
+
+        proc = subprocess.Popen(
+            [path] + shlex.split(request_json["calldata"]),
+            env=env,
```
Comment (sorawit) [inline cloud-function/google_cloud_function.py]: Don't think we need to pass env

File: cloud-function/google_cloud_function.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+            stderr=subprocess.PIPE,
+        )
+
+        proc.wait(timeout=timeout_sec)
+
+        returncode = proc.returncode
+        stdout = proc.stdout.read(MAX_STDOUT).decode()
+        stderr = proc.stderr.read(MAX_STDERR).decode()
+
+        return (
```
Comment (sorawit) [inline cloud-function/google_cloud_function.py]: Can we create a helper function `success(returncode="", stdout="", stderr="", err="")` so we don't repeat `jsonify(...), 200` everywhere?

**Review by sorawit — CHANGES_REQUESTED** [review]

File: CHANGELOG.md [hunk 4]
```diff
@@ -25,6 +25,7 @@
 
 ### Owasm
 
+- (feat) [\#1872](https://github.com/bandprotocol/bandchain/pull/1872) Add Owasm executor on Google Cloud function
```
Comment (taobun) [inline CHANGELOG.md]: I think this PR must be in chain not owasm or open new tag `executor`.

File: cloud-function/google_cloud_function.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+        return bad_request(err.messages)
+
+    path = "/tmp/execute.sh"
+    with open(path, "w") as f:
+        f.write(loaded_request.get("executable"))
+
+    os.chmod(path, 0o775)
+
+    try:
+
```
Comment (taobun) [inline cloud-function/google_cloud_function.py]: Remove endline

File: cloud-function/google_cloud_function.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+        returncode = proc.returncode
+        stdout = proc.stdout.read(MAX_STDOUT).decode()
+        stderr = proc.stderr.read(MAX_STDERR).decode()
+
+        return success(returncode, stdout, stderr, "")
+
+    except OSError:
+        return success(126, "", "", "Execution fail")
+
```
Comment (taobun) [inline cloud-function/google_cloud_function.py]: ditto

File: cloud-function/google_cloud_function.py [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+        )
+
+        proc.wait(timeout=timeout_sec)
+
+        returncode = proc.returncode
+        stdout = proc.stdout.read(MAX_STDOUT).decode()
+        stderr = proc.stderr.read(MAX_STDERR).decode()
+
+        return success(returncode, stdout, stderr, "")
+
```
Comment (taobun) [inline cloud-function/google_cloud_function.py]: ditto

File: cloud-function/main_test.py [hunk 8]
```diff
@@ -0,0 +1,386 @@
...
+        data=json.dumps({"executable": "123", "timeout": 123456}),
+        content_type="application/json",
+    )
+
+    data = json.loads(response.get_data(as_text=True))
+    assert response.status_code == 400
+    assert data["error"]["calldata"][0] == "field is missing from JSON request"
+
+
+def test_error_calldata_empty(mock_env):
```
Comment (taobun) [inline cloud-function/main_test.py]: Why this test is test_error?

File: cloud-function/main_test.py [hunk 9]
```diff
@@ -0,0 +1,386 @@
...
+    executable = encodeBase64(executable)
+    app = create_app()
+    response = app.test_client().post(
+        "/execute",
+        data=json.dumps(
+            {
+                "executable": executable,
+                "calldata": "bitcoin",
+                "timeout": "3000",
+            }  # timeout shouldn't be string
```
Comment (taobun) [inline cloud-function/main_test.py]: Remove comment

**Review by taobun — CHANGES_REQUESTED** [review]

**Review by taobun — APPROVED** [review]

File: cloud-function/google_cloud_function.py [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+            validate=validate.Range(min=0, max=MAX_TIMEOUT),
+            error_messages={"required": "field is missing from JSON request"},
+        )
+
+    try:
+        request_json = request.get_json(force=True)
+    except werkzeug.exceptions.BadRequest:
+        return bad_request("invalid JSON request format")
+
+    loaded_request = ""
```
Comment (sorawit) [inline cloud-function/google_cloud_function.py]: This is not needed.

File: cloud-function/google_cloud_function.py [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+    try:
+        loaded_request = RequestSchema().load(request_json)
+    except ValidationError as err:
+        return bad_request(err.messages)
+
+    path = "/tmp/execute.sh"
+    with open(path, "w") as f:
+        f.write(loaded_request.get("executable"))
```
Comment (sorawit) [inline cloud-function/google_cloud_function.py]: Can we do `loaded_request.executable`

**Review by sorawit — CHANGES_REQUESTED** [review]

**Review by sorawit — APPROVED** [review]

**Pull request merged**